### PR TITLE
SE-1822.Add.Decimal.Option

### PR DESCRIPTION
**Pull request by Omar-Oz**

### CODE REVIEW

File: python/generate/templates/api_client.mustache [hunk 1]
```diff
@@ -0,0 +1,731 @@
+# coding: utf-8
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:1]: Why is the default_templates file above empty? I have downloaded the file manually to compare against this file so we don't lose time while you add it.

File: python/generate/templates/api_client.mustache [hunk 2]
```diff
@@ -0,0 +1,731 @@
...
+        'str': str,
+        'bool': bool,
+        'date': datetime.date,
+        'datetime': datetime.datetime,
+        'object': object,
+    }
+    _pool = None
+
+    def __init__(self, configuration=None, header_name=None, header_value=None,
+                 cookie=None, pool_threads=1, flag_to_use_decimal= False):
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:65]: Remove the space between `flag_to_use_decimal=` and `False`. See from the PEP8 documentation (https://www.python.org/dev/peps/pep-0008/) 

_"Don't use spaces around the = sign when used to indicate a keyword argument, or when used to indicate a default value for an unannotated function parameter:"_

File: python/generate/templates/api_client.mustache [hunk 3]
```diff
@@ -0,0 +1,731 @@
...
+        self.configuration = configuration
+        self.pool_threads = pool_threads
+
+        self.rest_client = rest.RESTClientObject(configuration)
+        self.default_headers = {}
+        if header_name is not None:
+            self.default_headers[header_name] = header_value
+        self.cookie = cookie
+        # Set default User-Agent.
+        self.user_agent = '{{{httpUserAgent}}}{{^httpUserAgent}}OpenAPI-Generator/{{{packageVersion}}}/python{{/httpUserAgent}}'
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:77]: In the original this reads `{{#httpUserAgent}}{{{.}}}{{/httpUserAgent}}`. What is the significance of this change?

File: python/generate/templates/api_client.mustache [hunk 4]
```diff
@@ -0,0 +1,731 @@
...
+            if hasattr(atexit, 'unregister'):
+                atexit.unregister(self.close)
+
+    @property
+    def pool(self):
+        """Create thread pool on first request
+         avoids instantiating unused threadpool for blocking clients.
+        """
+        if self._pool is None:
+            atexit.registerF(self.close)
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:116]: This has been changed from `register` to `registerF`. Was that intentional?

File: python/generate/templates/api_client.mustache [hunk 5]
```diff
@@ -0,0 +1,731 @@
...
+        # save response body into a tmp file and return the instance
+        if response_type == "file":
+            return self.__deserialize_file(response)
+
+        # fetch data from response object
+
+
+        try:
+            if self.flag_to_use_decimal:
+                data = json.loads(response.data, parse_float = Decimal)
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:313]: Remove space surrounding the `=` sign as per PEP8.

File: python/generate/templates/api_client.mustache [hunk 6]
```diff
@@ -0,0 +1,731 @@
...
+            return 'application/json'
+
+        content_types = [x.lower() for x in content_types]
+
+        if 'application/json' in content_types or '*/*' in content_types:
+            return 'application/json'
+        else:
+            return content_types[0]
+
+    def update_params_for_auth(self, headers, queries, auth_settings,
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:574]: Why did you rename `querys` to `queries`. I understand that "querys" is not a word and that "queries" is the correct word to represent more than one query, however as it is outside the scope of the change here and we want to minimise the changes we make to the template as much as possible we should leave it as is.

File: python/generate/templates/api_client.mustache [hunk 7]
```diff
@@ -0,0 +1,731 @@
...
+                               request_auth=None):
+        """Updates header and query params based on authentication setting.
+
+        :param headers: Header parameters dict to be updated.
+        :param queries: Query parameters tuple list to be updated.
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:579]: `querys` renamed to `queries` here as well.

File: python/generate/templates/api_client.mustache [hunk 8]
```diff
@@ -0,0 +1,731 @@
...
+        :param auth_settings: Authentication setting identifiers list.
+        :param request_auth: if set, the provided settings will
+                             override the token in the configuration.
+        """
+        if not auth_settings:
+            return
+
+        if request_auth:
+            self._apply_auth_params(headers, queries, request_auth)
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:588]: And here

File: python/generate/templates/api_client.mustache [hunk 9]
```diff
@@ -0,0 +1,731 @@
...
+            return
+
+        for auth in auth_settings:
+            auth_setting = self.configuration.auth_settings().get(auth)
+            if auth_setting:
+                self._apply_auth_params(headers, queries, auth_setting)
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:594]: And here

File: python/generate/templates/api_client.mustache [hunk 10]
```diff
@@ -0,0 +1,731 @@
...
+
+    def _apply_auth_params(self, headers, queries, auth_setting):
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:596]: And here, are you sure that you used version 5.1.0 for the template? It seems unlikely that you made all these changes to the template.

Comment (MikeMcGarry) [reply python/generate/templates/api_client.mustache:596]: I won't highlight the rest, just do a diff with the template for version 5.1.0 yourself and ensure that we haven't changed any of these.

Comment (MikeMcGarry) [reply python/generate/templates/api_client.mustache:596]: https://github.com/OpenAPITools/openapi-generator/blob/v5.1.0/modules/openapi-generator/src/main/resources/python-legacy/api_client.mustache

File: python/generate/templates/api_client.mustache [hunk 11]
```diff
@@ -0,0 +1,731 @@
...
+        """Deserializes string to primitive type.
+
+        :param data: str.
+        :param klass: class literal.
+
+        :return: int, long, float, str, bool.
+        """
+
+        if type(klass) is Decimal and self.flag_to_use_decimal:
+            return Decimal(data)
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:648]: This should be inside the `try` / `except` in case a decimal can not be created.

Comment (MikeMcGarry) [reply python/generate/templates/api_client.mustache:648]: In fact on second thought, isn't this code redundant?

In the case that self.flag_to_use_decimal is set to `True` then the `NATIVE_TYPES_MAPPING` is updated so that `"float"` is resolved to `Decimal`. Therefore when we come through here the code below will just create a `Decimal`. 

Likewise if the flag is set to `False` then it will just create a `float`.

Comment (Omar-Oz) [reply python/generate/templates/api_client.mustache:648]: Ahh yes, you're right

File: python/generate/templates/api_client.mustache (hunk 1 above)
Comment (MikeMcGarry) [reply python/generate/templates/api_client.mustache:1]: Added, great.

File: python/generate/templates/api_client.mustache [hunk 12]
```diff
@@ -0,0 +1,725 @@
...
+        'str': str,
+        'bool': bool,
+        'date': datetime.date,
+        'datetime': datetime.datetime,
+        'object': object,
+    }
+    _pool = None
+
+    def __init__(self, configuration=None, header_name=None, header_value=None,
+                 cookie=None, pool_threads=1, flag_to_use_decimal=False):
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:65]: This is now redundant

File: python/generate/templates/api_client.mustache [hunk 13]
```diff
@@ -0,0 +1,725 @@
...
+
+        self.rest_client = rest.RESTClientObject(configuration)
+        self.default_headers = {}
+        if header_name is not None:
+            self.default_headers[header_name] = header_value
+        self.cookie = cookie
+        # Set default User-Agent.
+        self.user_agent = '{{#httpUserAgent}}{{{.}}}{{/httpUserAgent}}{{^httpUserAgent}}OpenAPI-Generator/{{{packageVersion}}}/python{{/httpUserAgent}}'
+        self.client_side_validation = configuration.client_side_validation
+        self.flag_to_use_decimal = flag_to_use_decimal
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:79]: Pull this off the configuration instead.

File: python/generate/templates/configuration.mustache [hunk 14]
```diff
@@ -177,7 +177,7 @@ conf = {{{packageName}}}.Configuration(
                  server_index=None, server_variables=None,
                  server_operation_index=None, server_operation_variables=None,
                  ssl_ca_cert=None,
-                 tcp_keep_alive=False,
+                 tcp_keep_alive=False,flag_to_use_decimal=False
```
Comment (MikeMcGarry) [inline python/generate/templates/configuration.mustache:180]: This should be on a new line.

Comment (MikeMcGarry) [reply python/generate/templates/configuration.mustache:180]: On second thought, I'll leave this up to your discretion, can't see anything in PEP8 guidelines about this.

File: python/generate/templates/rest.mustache [hunk 15]
```diff
@@ -15,6 +15,7 @@ import certifi
 import six
 from six.moves.urllib.parse import urlencode
 import urllib3
+import simplejson as json
```
Comment (MikeMcGarry) [inline python/generate/templates/rest.mustache:18]: Do we need to remove the `json` import above then? We have two imports with the same name here.

File: python/generate/templates/api_client.mustache [hunk 16]
```diff
@@ -0,0 +1,724 @@
+# coding: utf-8
+{{>partial_header}}
+from __future__ import absolute_import
+
+import atexit
+import datetime
+from dateutil.parser import parse
+import mimetypes
```
Comment (MikeMcGarry) [inline python/generate/templates/api_client.mustache:8]: Add the `json` import back here. Do you think we should use `simplejson` the same as we do in the rest.py file or the standard json library? I vote to stick with the standard if we don't need `simplejson` here and we can evaluate its effectiveness for deserialisation separately.

**Review by MikeMcGarry — APPROVED** [review]
